### PR TITLE
#76 PoC of junit reruns

### PR DESCRIPTION
**Pull request by Shaihuludus**

### CODE REVIEW

**Comment by youngeagle87** [comment]
approve

File: bb-aem-integration-tests/src/test/java/com/cognifide/bdd/demo/aem/wcm/SiteAdminTest.java [hunk 1]
```diff
@@ -28,6 +28,7 @@
 import java.util.Arrays;
 import java.util.List;
 
+import com.cognifide.qa.bb.qualifier.Retry;
```
Comment (mmanski) [inline bb-aem-integration-tests/src/test/java/com/cognifide/bdd/demo/aem/wcm/SiteAdminTest.java]: pls organize imports

File: bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java [hunk 2]
```diff
@@ -39,45 +39,40 @@
...
-
+  @Inject
+  private Injector injector;
   private final ThreadLocal<TestInfo> currentTest = new ThreadLocal<TestInfo>() {
     @Override
     protected TestInfo initialValue() {
       return injector.getInstance(TestInfo.class);
     }
   };
+  @Inject
```
Comment (mmanski) [inline bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java]: please add some separation here

File: bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java [hunk 3]
```diff
@@ -279,6 +274,14 @@ public Date getStartingDate() {
     return (Date) startingDate.clone();
   }
 
+  /**
```
Comment (mmanski) [inline bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java]: javadoc here could be ibherited from the interface here, this is redundant

File: bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java [hunk 4]
```diff
@@ -146,6 +158,39 @@ protected final void validateZeroArgConstructor(List<Throwable> errors) {
...
+  }
+
+  private int retrieveRetry(FrameworkMethod method) {
+    Retry retry = method.getAnnotation(Retry.class);
+    return retry != null ? calculateRetryCount(retry.reruns()) : 0;
+  }
+
+  private int calculateRetryCount(int reruns) {
+    return reruns > 0 ? reruns
+        : Integer.parseInt(properties.getProperty(ConfigKeys.JUNIT_RERUNS, "0"));
```
Comment (mmanski) [inline bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java]: the default here is zero but for `@Retry` is -1. Why such discrepancy? IMO -1 is somehow misleading and unclear for me, I cannot see the difference between 0 and -1 in the aspect of reruning tests

File: bb-aem-integration-tests/src/test/java/com/cognifide/bdd/demo/aem/wcm/SiteAdminTest.java (hunk 1 above)
Comment (Shaihuludus) [inline bb-aem-integration-tests/src/test/java/com/cognifide/bdd/demo/aem/wcm/SiteAdminTest.java]: fixed

File: bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java (hunk 2 above)
Comment (Shaihuludus) [inline bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java]: fixed

File: bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java (hunk 3 above)
Comment (Shaihuludus) [inline bobcat/bb-reports/src/main/java/com/cognifide/qa/bb/logging/TestEventCollectorImpl.java]: fixed

File: bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java (hunk 4 above)
Comment (Shaihuludus) [inline bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java]: fixed

File: bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java [hunk 5]
```diff
@@ -146,6 +158,39 @@ protected final void validateZeroArgConstructor(List<Throwable> errors) {
...
+        throw e;
+      } catch (Throwable e) {
+        if (runs >= retryCount) {
+          throw e;
+        }
+        testEventCollector.removeLastEntry();
+      }
+      runs++;
+    }
+    if(runs>1){
```
Comment (kkarolk) [inline bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java]: please fix formatting

**Comment by kkarolk** [comment]
approved but wait for travis

File: bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java (hunk 5 above)
Comment (Shaihuludus) [inline bobcat/bb-junit/src/main/java/com/cognifide/qa/bb/junit/TestRunner.java]: fixed

**Pull request merged**